Framework/NexusGeometry/src/JSONInstrumentBuilder.cpp

```diff
@@ -0,0 +1,141 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2019 ISIS Rutherford Appleton Laboratory UKRI,
+//     NScD Oak Ridge National Laboratory, European Spallation Source
+//     & Institut Laue - Langevin
+// SPDX - License - Identifier: GPL - 3.0 +
+
+#include "MantidNexusGeometry/JSONInstrumentBuilder.h"
+#include "MantidGeometry/Instrument.h"
+#include "MantidGeometry/Objects/IObject.h"
+#include "MantidNexusGeometry/InstrumentBuilder.h"
+#include "MantidNexusGeometry/NexusShapeFactory.h"
+#include "MantidNexusGeometry/TubeHelpers.h"
+
+namespace {
+using namespace Mantid::Geometry;
+using namespace Mantid::NexusGeometry;
+
+IObject_const_sptr createShape(const JSONGeometryParser &parser,
+                               const size_t bank) {
+  if (parser.isOffGeometry(bank))
+    return NexusShapeFactory::createFromOFFMesh(
+        parser.faces(bank), parser.windingOrder(bank), parser.vertices(bank));
+  else
+    return NexusShapeFactory::createCylinder(parser.cylinders(bank),
+                                             parser.vertices(bank));
+}
+
+IObject_const_sptr createMonitorShape(const Monitor &monitor) {
+  if (monitor.isOffGeometry)
+    return NexusShapeFactory::createFromOFFMesh(
+        monitor.faces, monitor.windingOrder, monitor.vertices);
+  else
+    return NexusShapeFactory::createCylinder(monitor.cylinders,
+                                             monitor.vertices);
+}
+
+Eigen::Matrix<double, 3, Eigen::Dynamic>
+getOffsets(const JSONGeometryParser &parser, const size_t index) {
+  // Initialise matrix
+  Eigen::Matrix<double, 3, Eigen::Dynamic> offsets;
+  int rowLength = 0;
+  const auto &x = parser.xPixelOffsets(index);
+  const auto &y = parser.yPixelOffsets(index);
+  const auto &z = parser.zPixelOffsets(index);
+
+  if (!x.empty())
+    rowLength = static_cast<int>(x.size());
+  else if (!y.empty())
+    rowLength = static_cast<int>(y.size());
+  // Need at least 2 dimensions to define points
+  else
+    return offsets;
+
+  // Default x,y,z to zero if no data provided
+  offsets.resize(3, rowLength);
+  offsets.setZero(3, rowLength);
+
+  if (!x.empty()) {
```

Comment:
Would it be more simple, and maybe even more performant, to construct `Eigen` matrices from `x`, `y`, `z` and then horizontally stack them with the `<<` syntax?

Reply:
What are your thoughts on this @OwenArnold ?

Reply:
Would be worth a try if it reduces code size.